### ISSUE TITLE
Metdata/Reflection on generics: TSC should emit the captured type

### ISSUE THREAD

**Issue by mindruptive** [Microsoft/TypeScript#7169]
TSC 175, target ES6, Node 5.6

Sample class with decorated properties:
```javascript
@sg.DMClass
export class User implements IUser {
    @sg.DMProperty
    firstName: string;
    @sg.DMProperty
    lastName: string;
    @sg.DMProperty
    friends: User[];
}
```
Property decorator:
```javascript
export function DMProperty(target: any, key: string) {
    var type = Reflect.getMetadata("design:type", target, key);
}
```

**Current behaviour:**
When the decorator gets called for property "friends", "type" is "Array" and the captured type ("User") is lost.

**Expected behaviour**
"type" should contain the captured type ("User"), too

Following very dirty work-around does the job: I modified tsc's method "emitSerializedTypeNode":
```javascript
function emitSerializedTypeNode(node) {
	if (node) {
		console.log("*** emitSerializedTypeNode: node.kind: ", node.kind);
		switch (node.kind) {
			...
			case 157:
				// Begin dirty ---------------------------------------------------------------------------
				var _elemType = node.elementType.typeName.text;
				write("{name: 'Array<" + _elemType + ">', type:'Array', elemType:'" + _elemType + "'}");
				console.log("*** emitSerializedTypeNode: Array, type: ", node.elementType.typeName.text);
				// End dirty -----------------------------------------------------------------------------
				return;
			case 150:
			...
```

Best
Mind

**Comment by mhegazy** [comment]
looks like a duplicate of https://github.com/Microsoft/TypeScript/issues/4521

**Comment by mindruptive** [comment]
Hi Mohamed,
sorry to bother you again. I don't think it's a duplicate of 4521. 4521 deals with the order of loading the modules, respectively circular refernces. That's not a problem in my case. All the modules are loaded correctly but as I can see from the compiler code (emitSerializedTypeNode), the captured type of a generic array is not emitted. Only "Array" gets emitted. So this behaviour is not (yet) in the compiler's code. 
Best
Mind

**Comment by mhegazy** [comment]
thanks. changing the title to reflect the request.

**Comment by mindruptive** [comment]
Thanks a lot!
Mind

**Comment by abierbaum** [comment]
Running into the same issue here.  Would be very helpful if we could get to the full nested type information.

**Comment by DanielTWalther** [comment]
This problem apparently still persists. And it really is a dealbreaker for one of my current projects. In an attempt to implement a workaround of my own, much like and inspired by @mindruptive , I created a [fork](https://github.com/INSPIRATIONlabs/TypeScript) of the TS Compiler, which currently achieves JS outputs along the lines of e.g.:
```
__metadata('design:type', {name: 'Array<Array<Number>>', type: Array, elemType: {name: 'Array<Number>', type: Array, elemType: Number}})
```
or
```
__metadata('design:type', {name: 'Array<InnermostTestClass>', type: Array, elemType: TestClass_1.InnermostTestClass})
```
for (nested) Arrays as well as lessen the condensing of UnionTypes like so:
```
__metadata('design:type', Boolean | Number | String)
```


I realize that this workaround, which is sufficient for said project, creates a breaking change to the behavior of metadata, as (e.g.) the return parameter of Reflect.getMetadata('design:type', target, key?) is no longer guaranteed to be a constructor.
My question (and request) now is: can/will this be resolved?

**Comment by DanielTWalther** [comment]
It seems as though there is actually some progress regarding this issue. From what I can tell after looking at the current TSC, there seems to be a modified implementation of the metadata decorators (and two separate instances of said decorator emitters (`function addOldTypeMetadata` and `function addNewTypeMetadata`); the new one (when enabled) currently emits javascript code like this:
```
__metadata("design:typeinfo", {
  type: () => Array
})
```
I'd suggest modifying the output to emit:
(e.g.: `number[][]`)
```
__metadata("design:typeinfo", {
  type: () => Array,
  elementType: () => {
    type: () => Array,
    elementType: () => Number
    }
})
```
This approach is obviously more elegant than building custom constructor functions (with or without a class or object mantle) to inject the elementType into.

**Comment by mhegazy** [comment]
I should note that Decorators have a [new proposal](https://github.com/tc39/proposal-decorators) in TC39 that should facilitate adding additional meta-data fields such as this one. I would expect TypeScript to implement the new proposal in the coming releases.

**Comment by DominicBoettger** [comment]
Are there any informations when design:typeinfo will be available?

**Comment by mhegazy** [comment]
> Are there any informations when design:typeinfo will be available?

No updates since https://github.com/Microsoft/TypeScript/issues/7169#issuecomment-250860376.

**Comment by abierbaum** [comment]
@mhegazy Do you know of any workaround that can be used in the meantime?  This is a major pain for a couple of our projects.

**Comment by DominicBoettger** [comment]
I think a workaround is only possible by defining the type in the decorator manually.

**Comment by thomas-mindruptive** [comment]
@DominicBoettger @mhegazy @abierbaum 
Since I still need this feature (I create a query-schema from TypeScript metadata) and I don't want to add unnecessary type info to my decorator, I patched the current version of tsc again:

```
function serializeTypeNode(node) {
	...
	case 163:
	case 164:
		// MR patch -------------------------------------
			// Patch comment out: return ts.createIdentifier("Array");
			console.log(">>> serializeTypeNode: " + node.kind + ", elementType.typeName.text: " + node.elementType.typeName.text );
			// Pass the name of the captured type to "createIdentifier()"
			return ts.createIdentifier("Array", node.elementType.typeName.text);
		// MR end patch --------------------------------   
}
    
// MR patch: add paramter "elemType"
function createIdentifier(text, elemType) {
	var node = createSynthesizedNode(70);
	node.text = ts.escapeIdentifier(text);
	// MR patch -----------------------
		if (elemType) {
			node.text = "{type:" + node.text + ", elemType:'" + elemType + "'}";
		}
	// MR end patch ------------------
	node.originalKeywordKind = text ? ts.stringToToken(text) : 0;
	node.autoGenerateKind = 0;
	node.autoGenerateId = 0;
	return node;
}
```

Finally to use the generated metadata in my property decorator:

```
function DMProperty(target: any, key: string)
	var type = Reflect.getMetadata("design:type", target, key);
	if (type.type && (type.type === "Array" || type.type.prototype instanceof Array || type.type === Array)) {
		// Do something with the type info, in our case: Save it to a schema property of the class.
		target[Schema].arrayTypes[key] = type.elemType;
	}
	...
}
```

From a compiler designer's perspective this might be dirty. TypeScript is great and the static type info makes any kind of schema-generation much easier. This slight enhancement even more so. :-)

**Comment by thomas-mindruptive** [comment]
```
   function serializeTypeNode(node) {
        ...
             case 164:
                    // MR Patch (case was empty before) *****************************************************
                    let typeNameText = undefined;
                    let typeName = node.elementType.typeName;
                    if (typeName) {
                        console.log(">>> serializeTypeNode: " + node.kind + ", elementType.typeName.text: " + typeName.text );
                        typeNameText = typeName.text;
                    } else {
                        let typeNameText = ts.tokenToString(node.elementType.kind);
                        console.log(">>> serializeTypeNode: " + node.kind + ", typeName is undefined => primitive type, using ts.tokenToString(): array elementType: " + typeNameText );
                    }
                    // Pass the name of the captured type to "createIdentifier()"
                    return ts.createIdentifier("Array", typeNameText);
                    // MR Patch end *************************************************************************
```

**Comment by danielweck** [comment]
+1 to avoid workarounds like https://github.com/edcarroll/ta-json#jsonelementtypetypefunction

**Comment by thomas-mindruptive** [comment]
and time goes by so slowly, ...

**Comment by yoxigen** [comment]
+1

**Comment by HideDev** [comment]
Hi, any news?

**Comment by inad9300** [comment]
Any progress after more than one year? Thanks :D

**Comment by thomas-mindruptive** [comment]
Excellent question! :-)

**Comment by JCWolf** [comment]
You may see low reaction numbers for this feature but it is waited by a lot of people. We need to transform JavaScript/Typescript and make it compete with languages like Java. 

For example, this feature will allow me to build a very beautiful database ORM just by using class properties and no any workarounds with decorators.

**Comment by NSExceptional** [comment]
@JCWolf Exactly. This would be super useful for ORMs. For example, if TypeORM were able to see the metadata type, you could change this code:

```ts
...
@OneToMany(type => Photo, "user")
photos: Photo[];
...
```
to this:

```ts
...
@OneToMany("user")
photos: Photo[];
...
```

I desperately want this feature.

**Comment by inad9300** [comment]
@NSExceptional The same goes for libraries like [TypedJSON](https://github.com/JohnWeisz/TypedJSON).

**Comment by danielweck** [comment]
Yes https://github.com/JohnWeisz/TypedJSON

...also https://github.com/edcarroll/ta-json

**Comment by thomas-mindruptive** [comment]
@JCWolf  I can provide the "patch-code" I have posted above. But I think it's more of a general discussion within the TS team if and how they want to do this. It would be easy (technically) for the transpiler to decorate additional type-info. Maybe they are waiting for the EcmaScript-specs for these features. (Which would be boring because TS was always ahead! :-))

**Comment by thomas-mindruptive** [comment]
And the years go by: > 3

**Comment by bemon** [comment]
+1

**Comment by BlueFrog130** [comment]
This would be really awesome... Been waiting on this feature awhile.

**Comment by maurei** [comment]
I feel like emitting generic type information is the next step towards having a strongly reflective language. At the same time I'm sure there must be technical/design dilemmas that haven't been covered in this thread. Either way, a comment from the typescript team here would be very welcome

**Comment by dolsem** [comment]
Any update on this? Would be nice to emit something like the following structure:
```ts
interface TypeMetadata {
  name: string;
  value: any;
  parameters: Array<TypeMetadata>;
}
```

So, for example, the result for `Map<string, MyInterface>` becomes:
```ts
{
  name: 'Map',
  value: Map,
  parameters: [
    { name: 'String', value: String, parameters: null },
    { name: 'MyInterface', value: Object, parameters: null },
  ],
}
```

**Comment by cuongphamcxa** [comment]
Any progress after more than one year?

**Comment by sshhsh** [comment]
Looking forward to updates. XD

**Comment by theoephraim** [comment]
+1

**Comment by comtaler** [comment]
I wish I am the last one to post this question: Any update?

**Comment by SpaceNinjaApe** [comment]
I've wrote a small example showing the use of decorators for request validations. Now i came across this issue.
I could do a dirty workaround like this:

example attr declaration.
```
@prop()
foo: bar;
```
check
```
if (Array.isArray(obj.foo)) { 
     obj.foo.forEach((elm) => {
            if (elm instanceof bar  === false) {
                return false
            }
     }
}
```

But then again i would rather prefer this:

example attr declaration.
```
@prop()
foo: bar[];
```
decorator-factory
```
function prop(): any {
    return function (target : Object, key : string) {
        let type = Reflect.getMetadata("design:type", target, key);
        console.log(type.isArray()) // true
        console.log(type.constructor.name) // "bar"
        PropertValidatorService.getInstance().createPropertyDescription(target, key, type.prototype, isPropertyEssential)
    }
}
```

**Comment by NSExceptional** [comment]
For everyone else who is, like me, wondering why this isn't getting any attention, a Redditor had this explanation for me:

> TypeScript's decorators are based on an older proposal, and the JS specs around decorators have gone in a different incompatible direction. I believe they're avoiding changing how decorators work until the final goal is specced. Another way of saying this is that you should avoid using decorators, because they are unlikely to see improvements any time soon.

With the disclaimer that this is just him reading the tea leaves.

**Comment by IonelLupu** [comment]
@NSExceptional That's a bit sad but I know what the TS team feels and they are right. It's so hard to do something that doesn't have a spec. 

I hope decorators will get specced soon because they are super awesome and  save you from writing a lot of code and makes it look really expressive.

**Comment by bemon** [comment]
Sometimes its better to ship implementation and made spec out of it based on real world needs. Decorators are super usefull and saves tons of code.

**Comment by thomas-mindruptive** [comment]
It's a pity. TypeScript used to lead the way and provide new functionality, which was not available in JS. 
Decorators have been around in TS for about 4 years. They are easy to use.

**Comment by ties-s** [comment]
Would love to see this happen.

**Comment by MaverikMinett** [comment]
Please give us this feature.

**Comment by Deathrage** [comment]
+1

**Comment by Limuyang1013** [comment]
Are their update on this?

**Comment by ionutnespus** [comment]
So, it's been seven years since initial request... Anyone found a solution for this?

**Comment by Yangff** [comment]
Implemented a transformer using ts-patch that can provide some workaround.. 
https://github.com/Yangff/extra-reflect-type
